src/coreclr/jit/llvm.cpp

```diff
@@ -1175,38 +1244,81 @@ void Llvm::buildCall(GenTree* node)
-    }
-    else if (cast->TypeIs(TYP_LONG) && genActualTypeIsInt(cast->CastOp()))
-    {
-        // Cast pointer to int if necessary.  TODO-LLVM: candidate for lowering?
-        Value* sourceValue = castIfNecessary(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastOp()->TypeGet()));
+    Value* castFromValue = consumeValue(cast->CastOp(), getLlvmTypeForVarType(castFromType));
+    Value* castValue = nullptr;
+    Type* castToLlvmType = getLlvmTypeForVarType(castToType);
 
-        mapGenTreeToValue(cast,
-            cast->IsUnsigned()
-            ? _builder.CreateZExt(sourceValue, getLlvmTypeForVarType(cast->CastToType()))
-            : _builder.CreateSExt(sourceValue, getLlvmTypeForVarType(cast->CastToType())));
-    }
-    else if (cast->TypeIs(TYP_INT, TYP_LONG) && cast->CastOp()->TypeIs(TYP_FLOAT, TYP_DOUBLE))
+    // TODO-LLVM: handle checked ("gtOverflow") casts.
+    switch (castFromType)
     {
-        mapGenTreeToValue(cast,
-            cast->IsUnsigned()
-                ? _builder.CreateFPToUI(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastToType()))
-                : _builder.CreateFPToSI(getGenTreeValue(cast->CastOp()), getLlvmTypeForVarType(cast->CastToType())));
-    }
-    else
-    {
-        // TODO: other casts
-        failFunctionCompilation();
+        case TYP_INT:
+        case TYP_LONG:
+            switch (castToType)
+            {
+                case TYP_BOOL:
+                case TYP_BYTE:
+                case TYP_UBYTE:
+                case TYP_SHORT:
+                case TYP_USHORT:
+                case TYP_INT:
+                case TYP_UINT:
+                    // "Cast(integer -> small type)" is "s/zext<int>(truncate<small type>)".
+                    // Here we will truncate and leave the extension for the user to consume.
+                    castValue = _builder.CreateTrunc(castFromValue, getLlvmTypeForVarType(castToType));
+                    break;
+
+                case TYP_LONG:
+                    castValue = cast->IsUnsigned()
+                                    ? _builder.CreateZExt(castFromValue, castToLlvmType)
+                                    : _builder.CreateSExt(castFromValue, castToLlvmType);
+                    break;
+
+                case TYP_DOUBLE:
+                    castValue = cast->IsUnsigned()
+                                    ? _builder.CreateUIToFP(castFromValue, castToLlvmType)
+                                    : _builder.CreateSIToFP(castFromValue, castToLlvmType);
+                    break;
+
+                default:
+                    failFunctionCompilation(); // NYI
+            }
+            break;
+
+        case TYP_FLOAT:
+        case TYP_DOUBLE:
+            switch (castToType)
+            {
+                case TYP_FLOAT:
+                case TYP_DOUBLE:
+                    castValue = _builder.CreateFPCast(castFromValue, getLlvmTypeForVarType(TYP_DOUBLE));
```

Comment:
```suggestion
                    castValue = _builder.CreateFPCast(castFromValue, castToLlvmType);
```
After this, you can delete the handling of FP types from `consumeValue`. All FP extensions and truncation should be explicit in the IR.